File: src/services/dataIssues.ts
/**
 * Client wrapper for the data-quality monitoring backend (ADR-012).
 *
 * Endpoints live on the payment-service (Litestar) — `Api.payment` axios
 * instance routes to `https://payment.transapp.ru/api/...` in prod and
 * to local docker dev URLs from `services/api.ts`.
 *
 * Used by:
 *   - `DataIssueReportButton`         → `reportDataIssue()`
 *   - `useSystemNotice`               → `fetchSystemNotice()`
 */

import { Platform } from 'react-native';

import Api from './api';
import type { ProviderId } from '../constants/providerLabels';

export type SystemNoticeSource = 'admin' | 'auto';

export interface SystemNoticeItem {
  category: ProviderId;
  message: string;
  source: SystemNoticeSource;
  /** ISO 8601 string from backend `created_at`. */
  since: string;
}

interface SystemNoticeResponse {
  notices: SystemNoticeItem[];
}

export interface ReportDataIssueInput {
  user_id: number;
  auto_id: number;
  category: ProviderId;
  comment?: string | null;
  fcm_token?: string | null;
}

interface ReportDataIssueResponse {
  id: number;
  notice_triggered: boolean;
}

/** Best-effort platform tag, matches backend Literal in
 *  `payment-service/app/schemas/data_issues.py`. */
function detectPlatform(): 'mobile_ios' | 'mobile_android' | 'web' {
  if (Platform.OS === 'ios') return 'mobile_ios';
  if (Platform.OS === 'android') return 'mobile_android';
  return 'web';
}

/**
 * POST a single complaint. Returns the issue id and whether it tripped
 * the auto-banner threshold (≥3 distinct user_ids in 6h for the same
 * category). Throws if the backend rejects (e.g. 409 — same user has
 * an open complaint for this auto/category).
 */
export async function reportDataIssue(
  input: ReportDataIssueInput,
): Promise<ReportDataIssueResponse> {
  const body = {
    user_id: input.user_id,
    auto_id: input.auto_id,
    category: input.category,
    comment: input.comment ?? null,
    fcm_token: input.fcm_token ?? null,
    platform: detectPlatform(),
  };
  const res = await Api.payment.post('/data-issues/report', body);
  return res.data as ReportDataIssueResponse;
}

/**
 * GET the currently active system notices. Polled by `useSystemNotice`;
 * cheap response (server-side it's a single SELECT on a partial-indexed
 * table). Returns `[]` when nothing is active.
 */
export async function fetchSystemNotice(): Promise<SystemNoticeItem[]> {
  const res = await Api.payment.get('/system-notice');
  const body = res.data as SystemNoticeResponse | null | undefined;
  return body?.notices ?? [];
}

File: payment-service/app/controllers/data_issues.py
"""
Litestar controller for client-facing data-issues endpoints.

POST /api/data-issues/report  — accept a single complaint
GET  /api/system-notice       — current active banners (polled by clients)
"""

from __future__ import annotations

import logging

from litestar import Controller, Request, get, post
from litestar.exceptions import HTTPException
from litestar.status_codes import (
    HTTP_201_CREATED,
    HTTP_409_CONFLICT,
)

from app.bot.notify import send_admin_alert
from app.schemas.data_issues import (
    ReportDataIssueRequest,
    ReportDataIssueResponse,
    SystemNoticeItem,
    SystemNoticeListResponse,
)
from app.services import data_issues as data_issues_svc
from app.services import system_notice as notice_svc

logger = logging.getLogger(__name__)


def _client_ip(request: Request) -> str | None:
    """Best-effort extraction of the originating client IP.

    nginx in front of payment-service is configured to forward the
    real address in `X-Forwarded-For`; the first value is the client.
    Falls back to the immediate peer address. Never raises.
    """
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        first = forwarded.split(",", 1)[0].strip()
        if first:
            return first[:45]  # IPv6 max
    if request.client and request.client.host:
        return request.client.host[:45]
    return None


class DataIssuesController(Controller):
    path = "/api"
    tags = ["data-issues"]

    @post("/data-issues/report", status_code=HTTP_201_CREATED)
    async def report(
        self, data: ReportDataIssueRequest, request: Request
    ) -> ReportDataIssueResponse:
        """Accept a single user complaint about a data category.

        Returns 409 if the user already has an open complaint for the
        same (auto, category) — partial UNIQUE INDEX in migration 002.

        Best-effort side effects (errors logged, never propagated):
          - Telegram admin alert
          - auto-banner threshold check
        """
        try:
            outcome = await data_issues_svc.report_issue(
                user_id=data.user_id,
                auto_id=data.auto_id,
                category=data.category,
                comment=data.comment,
                fcm_token=data.fcm_token,
                platform=data.platform,
                source_ip=_client_ip(request),
            )
        except data_issues_svc.IssueAlreadyOpenError as exc:
            raise HTTPException(
                status_code=HTTP_409_CONFLICT,
                detail="open complaint already exists for this user/auto/category",
            ) from exc

        # Fire-and-forget admin alert. We deliberately do NOT await within
        # a critical user-facing path that already returned a successful
        # DB write — but we DO await here because aiogram's HTTP call to
        # api.telegram.org is fast (~100ms p99). If that becomes a bottleneck
        # we move this to BackgroundTask.
        try:
            await send_admin_alert(
                issue_id=outcome.issue_id,
                user_id=data.user_id,
                auto_id=data.auto_id,
                category=data.category,
                comment=data.comment,
                auto_notice_triggered=outcome.notice_triggered,
                auto_notice_id=outcome.notice_id,
            )
        except Exception as exc:  # noqa: BLE001
            logger.warning(
                "send_admin_alert failed (non-fatal) issue_id=%d error=%s",
                outcome.issue_id,
                exc,
            )

        return ReportDataIssueResponse(
            id=outcome.issue_id,
            notice_triggered=outcome.notice_triggered,
        )

    @get("/system-notice")
    async def list_notices(self) -> SystemNoticeListResponse:
        """Return active system notices for the polling client."""
        notices = await notice_svc.get_active_notices()
        items = [
            SystemNoticeItem(
                category=n.category,  # type: ignore[arg-type]
                message=n.message,
                source=n.source,  # type: ignore[arg-type]
                since=n.created_at,
            )
            for n in notices
        ]
        return SystemNoticeListResponse(notices=items)
